src/Command/Auth/OAuth2Command.php

```diff
@@ -0,0 +1,202 @@
+                throw new \RuntimeException('File not found while waiting for authorization code: ' . $tmpFile);
+            }
+            $contents = file_get_contents($tmpFile);
+            if ($contents === false) {
+                unlink($tmpFile);
+                throw new \RuntimeException('Failed to read file while waiting for authorization code');
+            }
+            $code = trim($contents);
+        }
+
+        // Clean up.
+        // @todo make this more reliable
+        unlink($tmpFile);
+        unlink($listenerDir . '/index.php');
+        rmdir($listenerDir);
+
+        if (empty($code)) {
+            throw new \RuntimeException('Failed to get authorization code');
+        }
+
+        $this->stdErr->writeln('');
+        $this->stdErr->writeln('Login information received. Verifying...');
+
+        $this->saveAccessToken(
+            $this->getAccessToken($code, $localUrl)
+        );
+
+        /** @var \Doctrine\Common\Cache\CacheProvider $cache */
+        $cache = $this->getService('cache');
+        $cache->flushAll();
+
+        $info = $this->api()->getClient(false)->getAccountInfo();
+        if (isset($info['username'], $info['mail'])) {
+            $this->stdErr->writeln('');
+            $this->stdErr->writeln(sprintf(
+                'You are logged in as <info>%s</info> (%s).',
+                $info['username'],
+                $info['mail']
+            ));
+        }
+
+        return 0;
+    }
+
+    /**
+     * @param array $tokenData
+     */
+    private function saveAccessToken(array $tokenData)
+    {
+        $token = new AccessToken($tokenData['access_token'], $tokenData['token_type'], $tokenData);
+        $session = $this->api()->getClient(false)->getConnector()->getSession();
+        $session->setData([
+            'accessToken' => $token->getToken(),
+            'tokenType' => $token->getType(),
+        ]);
+        if ($token->getExpires()) {
+            $session->set('expires', $token->getExpires()->getTimestamp());
+        }
+        if ($token->getRefreshToken()) {
+            $session->set('refreshToken', $token->getRefreshToken()->getToken());
+        }
+        $session->save();
+    }
+
+    /**
+     * @param string $authCode
+     * @param string $redirectUri
+     *
+     * @return array
+     */
+    private function getAccessToken($authCode, $redirectUri)
+    {
+        return (new Client())->post(
+            $this->config()->get('service.accounts_url') . '/oauth2/token',
+            [
+                'json' => [
+                    'grant_type' => 'authorization_code',
+                    'code' => $authCode,
+                    'client_id' => $this->config()->get('api.oauth2_client_id'),
+                    'redirect_uri' => $redirectUri,
+                ],
+                'auth' => false,
+            ]
+        )->json();
+    }
+
+    /**
+     * @return string
+     */
+    private function getRandomState()
```

Comment:
Might be worth using `random_compat` instead of the boilerplate in there? Especially for crypto stuff...